presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/PushDownDereferences.java

```diff
@@ -0,0 +1,560 @@
+                new PushDownDereferenceThrough<>(RowNumberNode.class),
+                new PushDownDereferenceThrough<>(SortNode.class),
+                new PushDownDereferenceThrough<>(FilterNode.class),
+                new PushDownDereferenceThrough<>(LimitNode.class),
+                new PushDownDereferenceThroughProject(),
+                new PushDownDereferenceThroughUnnest(),
+                new PushDownDereferenceThroughSemiJoin(),
+                new PushDownDereferenceThroughJoin());
+    }
+
+    /**
+     * Extract dereferences and push them down to new ProjectNode below
+     * Transforms:
+     * <pre>
+     *  TargetNode(expression(a.x))
+     *  </pre>
+     * to:
+     * <pre>
+     *   ProjectNode(original symbols)
+     *    TargetNode(expression(symbol))
+     *      Project(symbol := a.x)
+     * </pre>
+     */
+    abstract class ExtractProjectDereferences<N extends PlanNode>
+            implements Rule<N>
+    {
+        private final Class<N> planNodeClass;
+
+        ExtractProjectDereferences(Class<N> planNodeClass)
+        {
+            this.planNodeClass = planNodeClass;
+        }
+
+        @Override
+        public Pattern<N> getPattern()
+        {
+            return Pattern.typeOf(planNodeClass);
+        }
+
+        @Override
+        public Result apply(N node, Captures captures, Context context)
+        {
+            Map<DereferenceExpression, VariableReferenceExpression> expressions =
+                    getDereferenceSymbolMap(ExpressionExtractor.extractExpressionsNonRecursive(node), context, metadata, sqlParser);
+
+            if (expressions.isEmpty()) {
+                return Result.empty();
+            }
+
+            return Result.ofPlanNode(new ProjectNode(context.getIdAllocator().getNextId(), rewrite(context, node, HashBiMap.create(expressions)), identityAssignmentsAsSymbolReferences(node.getOutputVariables())));
+        }
+
+        protected abstract N rewrite(Context context, N node, BiMap<DereferenceExpression, VariableReferenceExpression> expressions);
+    }
+
+    class ExtractFromFilter
+            extends ExtractProjectDereferences<FilterNode>
+    {
+        ExtractFromFilter()
+        {
+            super(FilterNode.class);
+        }
+
+        @Override
+        protected FilterNode rewrite(Context context, FilterNode node, BiMap<DereferenceExpression, VariableReferenceExpression> expressions)
+        {
+            PlanNode source = node.getSource();
+
+            ImmutableMap<VariableReferenceExpression, RowExpression>
+                    dereferencesMap =
+                    expressions.inverse().entrySet().stream().collect(toImmutableMap(Map.Entry::getKey,
+                            entry -> castToRowExpression(entry.getValue())));
+            Assignments assignments = Assignments.builder().putAll(identityAssignmentsAsSymbolReferences(source.getOutputVariables())).putAll(dereferencesMap).build();
```

Comment:
For this case:
Filter(predicate(a.x) = 5, output = a.y)
   Project(a)
-->
Filter(predicate(a_x) = 5, output = a_y)
   Project(a, deref(a.x) as a_x, deref(a.y) as a_y)

Is the new project correct, given that we are adding the existing project symbols and the deref expressions as new projects?